File: Sources/YToolsCore/SearchTextNormalizer.swift
import Foundation

public struct SearchTextForms: Equatable, Sendable {
    public let normalized: String
    public let abbreviation: String
    public let transliteration: String
    public let transliterationInitials: String
}

public struct SearchTextNormalizer: Sendable {
    private static let stableLocale = Locale(identifier: "en_US_POSIX")

    public init() {}

    public func forms(for value: String) -> SearchTextForms {
        let normalizedValue = normalized(value)
        let latin = value.applyingTransform(.toLatin, reverse: false) ?? value
        let latinWords = latin.split(whereSeparator: { !$0.isLetter && !$0.isNumber })
        let latinInitials = normalized(String(latinWords.compactMap(\.first)))
        let wordInitials = value
            .split(whereSeparator: { !$0.isLetter && !$0.isNumber })
            .compactMap(\.first)
        let nativeAbbreviation: String
        if wordInitials.count > 1 {
            nativeAbbreviation = normalized(String(wordInitials))
        } else {
            let characters = Array(value)
            nativeAbbreviation = normalized(String(characters.enumerated().compactMap { index, character in
                character.isUppercase || index == 0 ? character : nil
            }))
        }
        let containsNonASCII = value.unicodeScalars.contains { !$0.isASCII }

        return SearchTextForms(
            normalized: normalizedValue,
            abbreviation: containsNonASCII ? latinInitials : nativeAbbreviation,
            transliteration: normalized(latin),
            transliterationInitials: latinInitials
        )
    }

    public func normalized(_ value: String) -> String {
        value.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: Self.stableLocale)
            .filter { $0.isLetter || $0.isNumber }
    }

    /// Scores an ordered, non-contiguous match such as `slk` → `Slack`.
    /// Prefix and substring matches should still be ranked separately by callers.
    public func fuzzyScore(query: String, candidate: String) -> Int? {
        let queryCharacters = Array(normalized(query))
        let candidateCharacters = Array(normalized(candidate))
        guard !queryCharacters.isEmpty,
              queryCharacters.count <= candidateCharacters.count else { return nil }

        var searchStart = 0
        var previousMatch: Int?
        var firstMatch: Int?
        var score = 0

        for character in queryCharacters {
            guard let match = candidateCharacters[searchStart...].firstIndex(of: character) else {
                return nil
            }
            if firstMatch == nil { firstMatch = match }
            score += 12
            if let previousMatch {
                let gap = match - previousMatch - 1
                score += gap == 0 ? 8 : -min(gap * 2, 12)
            }
            previousMatch = match
            searchStart = match + 1
        }

        let coverage = queryCharacters.count * 30 / candidateCharacters.count
        let startPenalty = min((firstMatch ?? 0) * 2, 20)
        return max(1, min(99, score + coverage - startPenalty))
    }
}

File: src/YTools.Windows/Core/SearchTextNormalizer.cs
using System.Globalization;
using System.Text;
using TinyPinyin;

namespace YTools.Core;

public readonly record struct SearchTextForms(
    string Normalized,
    string Abbreviation,
    string Transliteration,
    string TransliterationInitials);

/// <summary>
/// Port of the macOS SearchTextNormalizer. Chinese text is transliterated to
/// pinyin with TinyPinyin; Latin text is case/diacritic folded with the
/// invariant culture so ranking is stable across system locales.
/// </summary>
public sealed class SearchTextNormalizer
{
    public SearchTextForms Forms(string value)
    {
        var normalizedValue = Normalized(value);
        var latin = ToLatin(value);
        var latinWords = SplitWords(latin);
        var latinInitials = Normalized(string.Concat(latinWords.Select(word => word[0])));
        var nativeWords = SplitWords(value);
        var wordInitials = nativeWords.Select(word => word[0]);

        string nativeAbbreviation;
        if (nativeWords.Count > 1)
        {
            nativeAbbreviation = Normalized(string.Concat(wordInitials));
        }
        else
        {
            nativeAbbreviation = Normalized(string.Concat(
                value.Select((character, index) => char.IsUpper(character) || index == 0 ? character.ToString() : "")));
        }

        var containsNonASCII = value.Any(character => character > 127);
        return new SearchTextForms(
            normalizedValue,
            containsNonASCII ? latinInitials : nativeAbbreviation,
            Normalized(latin),
            latinInitials);
    }

    public string Normalized(string value)
    {
        var decomposed = value.Normalize(NormalizationForm.FormD);
        var builder = new StringBuilder(decomposed.Length);
        foreach (var character in decomposed)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(character) == UnicodeCategory.NonSpacingMark)
            {
                continue;
            }

            if (char.IsLetterOrDigit(character))
            {
                builder.Append(char.ToLowerInvariant(character));
            }
        }

        return builder.ToString();
    }

    /// <summary>
    /// Scores an ordered, non-contiguous match such as `slk` → `Slack`.
    /// Prefix and substring matches should still be ranked separately by callers.
    /// </summary>
    public int? FuzzyScore(string query, string candidate)
    {
        var queryCharacters = Normalized(query).ToCharArray();
        var candidateCharacters = Normalized(candidate).ToCharArray();
        if (queryCharacters.Length == 0 || queryCharacters.Length > candidateCharacters.Length)
        {
            return null;
        }

        var searchStart = 0;
        int? previousMatch = null;
        int? firstMatch = null;
        var score = 0;

        foreach (var character in queryCharacters)
        {
            var match = Array.IndexOf(candidateCharacters, character, searchStart);
            if (match < 0)
            {
                return null;
            }

            firstMatch ??= match;
            score += 12;
            if (previousMatch is { } previous)
            {
                var gap = match - previous - 1;
                score += gap == 0 ? 8 : -Math.Min(gap * 2, 12);
            }

            previousMatch = match;
            searchStart = match + 1;
        }

        var coverage = queryCharacters.Length * 30 / candidateCharacters.Length;
        var startPenalty = Math.Min((firstMatch ?? 0) * 2, 20);
        return Math.Max(1, Math.Min(99, score + coverage - startPenalty));
    }

    private static List<string> SplitWords(string value)
    {
        var words = new List<string>();
        var current = new StringBuilder();
        foreach (var character in value)
        {
            if (char.IsLetterOrDigit(character))
            {
                current.Append(character);
            }
            else if (current.Length > 0)
            {
                words.Add(current.ToString());
                current.Clear();
            }
        }

        if (current.Length > 0)
        {
            words.Add(current.ToString());
        }

        return words;
    }

    private static string ToLatin(string value)
    {
        var builder = new StringBuilder(value.Length + 8);
        foreach (var character in value)
        {
            if (character <= 127)
            {
                if (char.IsLetterOrDigit(character))
                {
                    builder.Append(character);
                }
                else
                {
                    builder.Append(' ');
                }

                continue;
            }

            if (PinyinHelper.IsChinese(character))
            {
                var pinyin = PinyinHelper.GetPinyin(character);
                if (!string.IsNullOrEmpty(pinyin))
                {
                    builder.Append(pinyin);
                    builder.Append(' ');
                }
                else
                {
                    builder.Append(' ');
                }

                continue;
            }

            builder.Append(char.IsLetterOrDigit(character) ? character : ' ');
        }

        return builder.ToString();
    }
}
